source/iterator/snapshot.go

```diff
@@ -0,0 +1,234 @@
+		IncludedFields: query.ExecuteOptions_ALL,
+	})
+
+	if params.Position != nil {
+		snapshot.position = params.Position
+	}
+
+	return snapshot, nil
+}
+
+// hasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (s *snapshot) hasNext(ctx context.Context) (bool, error) {
+	if len(s.records) == 0 {
+		if err := s.loadRecords(ctx); err != nil {
+			return false, fmt.Errorf("load records: %w", err)
+		}
+	}
+
+	return len(s.records) > 0, nil
+}
+
+// next returns the next record.
+func (s *snapshot) next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+
+	case record := <-s.records:
+		return record, nil
+	}
+}
+
+// stop does nothing.
+func (s *snapshot) stop(ctx context.Context) error {
+	sdk.Logger(ctx).Debug().Msgf("stop snapshot iterator")
+
+	return nil
+}
+
+// loadRecords selects a batch of rows from a database, based on the snapshot's
+// table, columns, orderingColumn, batchSize and the current position,
+// and converts them to the sdk.Record.
+func (s *snapshot) loadRecords(ctx context.Context) error {
+	selectDataset := goqu.Dialect("mysql").Select()
+
+	if len(s.columns) > 0 {
+		keyColumnPresent := false
+		// add one to the capacity to have a space for the keyColumn
+		// if it's not present in the columns list.
+		cols := make([]any, 0, len(s.columns)+1)
+		for i := 0; i < len(s.columns); i++ {
+			if s.columns[i] == s.keyColumn {
+				keyColumnPresent = true
+			}
+
+			cols = append(cols, s.columns[i])
+		}
+
+		if !keyColumnPresent {
+			cols = append(cols, s.keyColumn)
+		}
+
+		selectDataset = selectDataset.Select(cols...)
+	}
+
+	selectDataset = selectDataset.
+		From(s.table).
+		Order(goqu.C(s.orderingColumn).Asc()).
+		Limit(uint(s.batchSize))
+
+	if s.position != nil {
+		selectDataset = selectDataset.Where(
+			goqu.Ex{
+				s.orderingColumn: goqu.Op{"gt": s.position.LastProcessedElementValue},
+			},
+		)
+	}
+
+	query, _, err := selectDataset.ToSQL()
+	if err != nil {
+		return fmt.Errorf("construct select query: %w", err)
+	}
+
+	resultStream, err := s.session.StreamExecute(ctx, query, nil)
+	if err != nil {
+		return fmt.Errorf("stream execute select query: %w", err)
+	}
+
+	if err := s.processStreamResults(ctx, resultStream); err != nil {
+		return fmt.Errorf("process stream results: %w", err)
+	}
+
+	return nil
+}
+
+// processStreamResults receives a result from the sqltypes.ResultStream,
+// converts its rows to the sdk.Record and sends them to the i.records channel.
+func (s *snapshot) processStreamResults(ctx context.Context, resultStream sqltypes.ResultStream) error {
+	for {
+		result, err := resultStream.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+
+			return fmt.Errorf("stream recv: %w", err)
+		}
+
+		if result.Fields != nil {
+			s.fields = result.Fields
+		}
+
+		for _, row := range result.Rows {
+			if s.fields == nil {
```

Comment:
Why not use `result.Fields` directly?

Reply:
Yeah, you're right. We don't need the `snapshot.fields` in the snapshot.

Reply:
Actually, we need to store this in a separate variable because the `result.Fields` may be nil for some results.